Code/src/main/java/nl/utwente/group10/ui/components/blocks/FunctionBlock.java

```diff
@@ -156,20 +169,173 @@ public final int getArgumentIndex(ConnectionAnchor anchor) {
     }
 
     @Override
+    public boolean inputsAreConnected() {
+        for (int i = 0; i < getInputs().length; i++) {
+            if (!inputIsConnected(i)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean inputIsConnected(int index) {
+        if (index >= 0 && index < getInputs().length) {
+            return getInputs()[index] != null
+                    && getInputs()[index].isFullyConnected();
+        }
+        return false;
+    }
+
+    /**
+     * @return The current (output) expression belonging to this block.
+     */
+    @Override
     public final Expr asExpr() {
         Expr expr = new Ident(getName());
-        for (InputAnchor in : getInputs()) expr = new Apply(expr, in.asExpr());
-
+        // Find last input that is filled in
+        for (InputAnchor in : getInputs()) {
+            if (in != null) {
+                Expr inputExpr = in.asExpr();
+                expr = new Apply(expr, inputExpr);
+            }
+        }
         return expr;
     }
 
+    /*
+     * Signature = non unified type, ie: a->b
+     * 
+     * (Current)Type = unified type, ie Int -> Float (This can still have
+     * signature a->b)
+     * 
+     * These are not the same, but are related. The Type has to conform to the
+     * signature.
+     */
+
+    public Type getFunctionSignature() {
+        return getFunctionSignature(getPane().getEnvInstance(), new GenSet());
+    }
+
+    public Type getFunctionSignature(Env env, GenSet genSet) {
+        return env.getFreshExprType(this.getName()).get();
+    }
+
+    @Override
+    public Type getOutputSignature() {
+        return getOutputSignature(getPane().getEnvInstance(), new GenSet());
+    }
+
+    @Override
+    public Type getOutputSignature(Env env, GenSet genSet) {
+        Type type = getFunctionSignature();
+        for (int i = 0; i < getInputs().length; i++) {
+            type = ((ConstT) type).getArgs()[1];
```

Comment:
Mits de functie correct gedefinieerd is in de catalogus gaat dit goed, ik zal een exception toevoegen